src/extensions/extension_manager/src/extension_manager_modules.cpp

```diff
@@ -0,0 +1,237 @@
+        result.ExtendedResultCode = ADUC_ERC_EXTENSION_CREATE_FAILURE_NOT_FOUND(facilityCode, componentCode);
+        goto done;
+    }
+
+    // Validate file hash.
+    if (!ADUC_HashUtils_GetShaVersionForTypeString(
+            ADUC_HashUtils_GetHashType(entity.Hash, entity.HashCount, 0), &algVersion))
+    {
+        Log_Error(
+            "FileEntity for %s has unsupported hash type %s",
+            entity.TargetFilename,
+            ADUC_HashUtils_GetHashType(entity.Hash, entity.HashCount, 0));
+        result.ExtendedResultCode = ADUC_ERC_EXTENSION_CREATE_FAILURE_VALIDATE(facilityCode, componentCode);
+        goto done;
+    }
+
+    if (!ADUC_HashUtils_IsValidFileHash(
+            entity.TargetFilename,
+            ADUC_HashUtils_GetHashValue(entity.Hash, entity.HashCount, 0),
+            algVersion,
+            true /* suppressErrorLog */))
+    {
+        Log_Error("Hash for %s is not valid", entity.TargetFilename);
+        result.ExtendedResultCode = ADUC_ERC_EXTENSION_CREATE_FAILURE_VALIDATE(facilityCode, componentCode);
+        goto done;
+    }
+
+    *libHandle = dlopen(entity.TargetFilename, RTLD_LAZY);
+
+    if (*libHandle == nullptr)
+    {
+        Log_Error("Cannot load handler file %s. %s.", entity.TargetFilename, dlerror());
+        result.ExtendedResultCode = ADUC_ERC_EXTENSION_CREATE_FAILURE_LOAD(facilityCode, componentCode);
+        goto done;
+    }
+
+    dlerror(); // Clear any existing error
+
+    // Only check whether required function exist, if specified.
+    if (requiredFunction != nullptr && *requiredFunction != '\0')
+    {
+        void* reqFunc = dlsym(*libHandle, requiredFunction);
+
+        if (reqFunc == nullptr)
+        {
+            Log_Error("The specified function ('%s') doesn't exist. %s\n", requiredFunction, dlerror());
+            result.ExtendedResultCode =
+                ADUC_ERC_EXTENSION_FAILURE_REQUIRED_FUNCTION_NOTIMPL(facilityCode, componentCode);
+            goto done;
+        }
+    }
+
+    // Cache the loaded library.
+    ExtensionManager::_libs.emplace(extensionName, *libHandle);
+
+    result = { ADUC_Result_Success };
+
+done:
+    if (IsAducResultCodeFailure(result.ResultCode))
+    {
+        if (*libHandle != nullptr)
+        {
+            dlclose(*libHandle);
+            *libHandle = nullptr;
+        }
+    }
+
+    // Done with file entity.
+    ADUC_FileEntity_Uninit(&entity);
+
+    return result;
+}
+
+ADUC_Result ExtensionManager::LoadAgentModules()
+{
+    ADUC_Result result{ ADUC_GeneralResult_Failure };
+
+    struct dirent *entry;
+    DIR *dir = opendir(ADUC_AGENT_MODULE_REGISTRY_FOLDER);
+
+    if (dir == NULL) {
+        Log_Error("Error opening directory");
+        // TODO: Set ADUC_Result
+        goto done;
+    }
+
+    char *folders[ADUC_MAX_AGENT_MODULE];  // Assuming max 100 extensions for simplicity
+    int count = 0;
+
+    while ((entry = readdir(dir)) != NULL) {
```

Comment:
needs to be ADUCPAL_readdir for cross-platform, but recommend using ADUCPAL_scandir and then looping through the folder names it outputs.